saleor/graphql/product/types/products.py

```diff
@@ -784,15 +785,23 @@ def resolve_variants(root: ChannelContext[models.Product], info, **_kwargs):
         is_staff = requestor_is_staff_member_or_app(requestor)
 
         def return_available_variants(variants):
-            if is_staff:
+            if is_staff and not root.channel_slug:
                 return variants
 
             variants_id = (variant.id for variant in variants)
-            available_variants_in_channel = (
-                AvailableProductVariantsByProductVariantIdAndChannel(info.context).load(
-                    (variants_id, root.channel_slug)
+            if is_staff and root.channel_slug:
+                available_variants_in_channel = (
+                    ProductVariantsByProductVariantIdAndChannel(info.context).load(
```

Comment:
We shouldn't fetch `ProductVariants` from the database 2 times.  
We should create `ProductVariantsByProductIdAndChannel` and `AvailableProductVariantsByProductIdAndChannel` and refactor whole resolver.